remark-lint-nodejs-links.js

```diff
@@ -21,7 +21,7 @@ function validateLinks(tree, vfile) {
   for (const node of getLinksRecursively(tree)) {
     if (node.url[0] !== "#") {
       const targetURL = new URL(node.url, currentFileURL);
-      if (targetURL.protocol === "file:" && !fs.existsSync(targetURL)) {
```

Comment:
Another problem is: We're usually referring the relative paths in our md file for each other, for example in the node.org project, we have a reference like this:

![image](https://user-images.githubusercontent.com/52018749/126019280-6acf007a-5af8-4e72-8bec-6ec082ed2389.png)

This link, will be generated to "http" by build.js in nodejs.org project, however it has a "file:///" as the URL at local disc, this means in remark-lint-nodejs-links.js, the actual path of the file would be:

![image](https://user-images.githubusercontent.com/52018749/126019372-5fc3a1d6-ef05-4cc7-b6a2-132f145d1e8c.png)

Of course, it doens't exist, but it EXISTS really when deploying onto the server with an "http://".

Reply:
On nodejs/node repo, we forbid any path that starts with a `/`, we expect every path to be relative.
If nodejs/nodejs.org uses absolute paths, the way forward would be to add an option to specify what is the absolute base. The implementation would look something like that:


```js
const targetURL = (options.absoluteBase != null && node.url[0] === "/") ?
  new URL('.' + node.url, options.absoluteBase) :
  new URL(node.url, currentFileURL);
```
  

> This experiment is done locally on my PC of folked node.org project, however, the pdf files really exist:

I suspect the `static` folder you show in the screen shot is not at `D:\static`, right? Instead, I suspect it is located at `D:\Projects\node_proj\nodejs.org\static` or something like that.

Reply:
>I suspect the static folder you show in the screen shot is not at D:\static, right? Instead, I suspect it is located at D:\Projects\node_proj\nodejs.org\static or something like that.

@aduh95: 
Yes, that's the big problem.

BTW：I'm not sure how many kinds of paths we have now in these *.md files, sometimes we will have a path with a "#", pointing at a certain pharagraphu in another md file, and sometimes we have something like :/en/about, but it points at /en/about/index.md....

Too complicated to think...

Reply:
I may be missing some, but I think there are only 5 types of URL:

* Fully specified URL, e.g. `https://nodejs.org/en/#anchor` (currently supported, ignored if using a protocol other than `file:`).
* Protocol relative URL, e.g. `//nodejs.org/en/#anchor` (not supported, but also not used AKAICT).
* Origin-relative URL, e.g. `/en/#anchor` (not supported, that's the issue you are referring to).
* Relative URL, e.g. `./en/#anchor` (supported).
* Document-relative URL, e.g. `#anchor` (supported).

If we want to support origin-relative URLs, we need to either:
- try to resolve them from a root folder (provided by the configuration or the environment).
- ignore them, so not try to find out if it targets a file that exists or not.

In either case, it needs to come from a configuration, as we wouldn't want to have an origin-relative URL in the nodejs/node repo.

> sometimes we have something like :/en/about, but it points at /en/about/index.md....

I don't think that's going to be a problem as long as `en/about` exists on the file system (it doesn't matter if it's a directory or a markdown file).

Reply:
Thanks for your suggestions in detail, and after deep thought for a while, I don't it's good to mix your corrent checking with this specific kind of md checks. What's more, considering there's a problem of ASCII ordering and I've disabled the check. So maybe I'll consider making another tool or package from npm instead of directly writing in your package :)